posts/2022-10-04-microprofile-serverless-ibm-code-engine.adoc

```diff
@@ -0,0 +1,414 @@
+
+If you are missing any required properties, when your application attempts to build the CloudEvent an exception will be thrown informing of a missing property. To see which properties are required you can review the https://github.com/cloudevents/spec/blob/v1.0/spec.md[specification]
+
+As CloudEvents can be from a wide variety of sources can that differ even within a single provider the majority of fields are fairly free-form.
+
+==== Receiving a CloudEvent
+Having returned a CloudEvent, how can one be handled within the application.
+
+For this we are going to enhance the `SystemResource` class to add a query method where we can send a POST request where the body contains the system properties that we want returned.
+
+The body we are going to send will be a JSON array with each property we want as an a member of the array
+```
+["java.vendor.url","awt.toolkit"]
+```
+
+In the class add the following Method declaration.
+```
+    @POST
+    @Produces(MediaType.APPLICATION_JSON)
+    @Timed(name = "queryPropertiesTime",
+            description = "Time needed to query the JVM system properties")
+    @Counted(absolute = true, description
+            = "Number of times the JVM system properties are queried")
+    public CloudEvent queryProperties(CloudEvent query){
+
+
+}
+```
+In this case we will be returning a CloudEvent, but you can return any type that you want that confirms the request was received such as `Response.ok().build();`
+
+Inside the method, we need to perform some similar actions as we did within `getProperties()`, but also handle the CloudEvent input.
+
+As some initial structure add this block to the top of the method
+```
+Map properties = System.getProperties();
+HashMap<String,String> props = new HashMap<>((Map<String,String>)properties);
+HashMap<String,String> qProps = new HashMap<String,String>();
+Jsonb jsonb = JsonbBuilder.create();
+```
+This gives us the Map of the properties, but in a form that is more useful later as we need to do more processing then we did within `getProperties()`
+
+To retrieve the data from the CloudEvent we use `.withData();` to extra and Object of type CloudEventData, the data portion is in binary format, so needs to be converted to make it usable.
+```
+/* Pull data from the data portion of the CloudEvent - this is in binary format so convert it into a standard String*/
+CloudEventData data = query.getData();
+String jsonString = new String(data.toBytes(), StandardCharsets.UTF_8);
+```
+The conversion to a String is the quickest way if you know the input will be in format of JSON, you can check what the sender claims it has sent you can validate via inspect the Data Content Type from `getDataContentType()`
+
+With the data now in a more usable format we can start to process it and make use of its contents.
+
+As we have a JSON array, we can use jsonb to convert the json to an ArrayList of the keys requested from the properties HashMap.
```

Comment:
```suggestion
Because we have a JSON array, we can use `jsonb` to convert the JSON to an `ArrayList` of the keys that are requested from the properties `HashMap`.
```